File: cme.py
import numpy as np

from tqdm import tqdm

import logging
logger = logging.getLogger(__name__)

try:
    import pdist
except ModuleNotFoundError:
    logger.warning("Could not load module pdist")
    
dtype = np.double

class Reaction:
    """ Reaction base class 

        Every reaction has the following parameters:

        - rate: float
            Base rate of the reaction

        - dist: float (default: 0)
            Distance from epicenter at which products appear

        - products: array or None (default: None)
            List of products created during the reaction. Entries must be of the following forms:
            * spec: int 
                Denotes one reactant of the given species 
            * (b: float, spec: int)
                Denotes a burst of the given species following the geometric distribution with mean b

    """

    def __init__(self, rate, dist = 0, products = None):
        self.rate = rate
        self.dist = dist
        self.products = products

class GenReaction(Reaction):
    """
        The GenReaction class represents reactions without educts.
    """

    def __init__(self, rate, dist = 0, products = None):
        super().__init__(rate, dist, products)
        
    def __str__(self):
        return "GenReaction(rate={}, dist={}, products={})".format(self.rate, self.dist, self.products)
        
class UniReaction(Reaction):
    """ The UniReaction class represents unimolecular reactions. 
        
        Parameters:
        - spec: int
            The species undergoing the reaction
    """

    def __init__(self, rate, spec, dist = 0, products = None):
        super().__init__(rate, dist, products)
        self.spec = spec
        
    def __str__(self):
        return "UniReaction(rate={}, spec={}, dist={}, products={})".format(self.rate, self.spec, self.dist, self.products)
        
class BiReaction(Reaction):
    """
        The BiReaction class represents bimolecular reactions. Note that for reactions 
        of two particles of the same species every combination of two particles is only counted once.

        Parameters:
        - specA, specB: int
            The species of the two particles undergoing the reaction
    """

    def __init__(self, rate, specA, specB, dist = 0, products = None):
        super().__init__(rate, dist, products)
        self.specA = specA
        self.specB = specB
        
    def __str__(self):
        return "BiReaction(rate={}, specA={}, specB={}, dist={}, products={})".format(self.rate, self.specA, self.specB, self.dist, self.products)
    
class ReactionSystem:
    """ This class stores information about the model in a CME reaction system.
    
        Arguments:
            n_species: int
                Number of species in the system. Species are 
                enumerated starting from 0.
               
            reactions: array of reactions (default: ())
                List of allowed reactions in the system
            
            initial_state: array of ints
                Initial number of particles per species. 
                
        Attributes:
            block_size: int (default: 100)
                Size of block in particle buffer (internal)
    """
    def __init__(self, n_species, reactions = (), initial_state = None):
        self.n_species = n_species
        
        if initial_state is None:
            initial_state = np.zeros(n_species, dtype=int)
        
        self.initial_state = initial_state
        
        self.reactions = reactions
        self.sort_reactions(reactions)
        
    def sort_reactions(self, reactions):
        """ Sort reactions by type during initialisation """
        self.reactions_gen = []
        self.reactions_uni = []
        self.reactions_bi = []
        
        for reac in reactions:
            if isinstance(reac, GenReaction):
                self.reactions_gen.append(reac)
            elif isinstance(reac, UniReaction):
                self.reactions_uni.append(reac)
            elif isinstance(reac, BiReaction):
                self.reactions_bi.append(reac)
            else:
                raise TypeError("Unknown reaction type for '{}'".format(reac))
        
    def create_particle_system(self, seed=None):
        """ Create a new instance of ParticleSystem associated to this reaction system """
        return ParticleSystem(self, seed=seed)
    
class ParticleSystem:
    """ This class represents the state of a particle system associated to a ReactionSystem 
    
        Args:
            system: ReactionSystem
                Associated reaction system defining the model to use
                
            seed: int or None (default: None)
                Optional seed for random number generator
                
        Attributes:
            counts: n_species array of ints
                Per-species particle counts
                
            t: float
                Current time in the system
                
            events: array of events
                Ordered list of time-stamped events in the system
    """
    def __init__(self, system, seed=None):
        self.system = system
        
        self.counts = np.zeros(system.n_species, dtype=int)
        
        self.rng = np.random.RandomState(seed=seed)
       
        # Precompute some things for ease of access during simulations
        self.gen_rates = np.asarray([ r.rate for r in self.system.reactions_gen ])
        
        self.uni_rates = np.asarray([ r.rate for r in self.system.reactions_uni ])
        self.uni_spec = np.asarray([ r.spec for r in self.system.reactions_uni ], dtype=int)
       
        self.t = 0
        self.events = []
        
        # Used for fast updating of particle count trajectories 
        self.events_iter = None
        
        self.add_initial_molecules()
        
    ### INTERNALS ###
    def add_products(self, products):
        """ Add list of particles to the reaction system. Returns list of particles placed.
        """

        # Create list of products to be placed
        products = self.expand_products(products)
        
        log = []
        for product in products:
            self.counts[product] += 1
            
            log.append((product,))
            
        return log
    
    def expand_products(self, products):
        """ 
            Convert description of reaction products into list of products to be
            added, drawing particle numbers for products produced in bursts.
        """
        ret = []
        for prod in products:
            if type(prod) == int:
                ret.append(prod)
            else:
                m, spec = prod
                
                p = 1 / m
                n = self.rng.geometric(p) - 1
                ret += [ spec for i in range(n) ]
                
        return ret
    
    def add_initial_molecules(self):
        """ Place initial molecules in the system """
        assert self.t == 0
        
        for spec, n_init in enumerate(self.system.initial_state):
            product_log = self.add_products([ spec for i in range(n_init) ])
                
            event = ("gen", None, product_log)
            self.events.append((0, event))
            
    def compute_bi_rates(self):
        """ Compute reaction rates for bimolecular reactions """
        rates = np.empty(len(self.system.reactions_bi),)
        
        for i, reac in enumerate(self.system.reactions_bi):
            # Do not possible overcount reactant pairs if both educts are of the same species
            if reac.specA == reac.specB:
                combs = 0.5 * self.counts[reac.specA] * (self.counts[reac.specA] - 1)
            else:
                combs = self.counts[reac.specA] * self.counts[reac.specB]
                
            rates[i] = reac.rate * combs
            
        return rates
    
    ### UPDATES ###
    def perform_gen_reaction(self, reaction):
        """ Simulate one occurrence of the specified generative reaction """
        product_log = self.add_products(reaction.products)
        return ("gen", reaction, product_log)
                        
    def perform_uni_reaction(self, reaction):
        """ Simulate one occurrence of the specified unimolecular reaction """
        self.counts[reaction.spec] -= 1
        product_log = ()
        
        if reaction.products is not None:
            product_log = self.add_products(reaction.products)
            
        return ("uni", reaction, product_log)
    
    def perform_bi_reaction(self, reaction, rates):
        """ Simulate one occurrence of the specified bimolecular reaction.
            Refer to perform_uni_reaction for more information. """
        
        self.counts[reaction.specA] -= 1
        self.counts[reaction.specB] -= 1
        
        product_log = ()
        if reaction.products is not None:
            product_log = self.add_products(reaction.products)
            
        return ("bi", reaction, product_log)
        
    ### MAIN LOOP ###
    def run(self, tmax, disable_pbar=True):
        """ Run simulation for tmax time units using the Gillespie algorithm """
        t0 = self.t
        
        gen_rates = self.gen_rates
        
        with tqdm(total=tmax, 
                  desc="Time simulated: ", 
                  unit="tu", 
                  disable=disable_pbar) as pbar:
            while True:
                uni_rates = self.uni_rates * self.counts[self.uni_spec]
                bi_rates = self.compute_bi_rates()
                rate = np.sum(gen_rates) + np.sum(uni_rates) + np.sum(bi_rates)
                
                # Nothing happening
                if rate == 0.0 or not np.isfinite(rate):
                    if not np.isfinite(rate):
                        logger.warning("Numerical blow-up in CME simulation")
                       
                    # Pretend the last reaction happened at time tmax.
                    # This is necessary for updating progress bar correctly.
                    dt = 0
                    break

                dt = self.rng.exponential(1 / rate)
                self.t += dt
                if self.t > t0 + tmax:
                    break
                    
                pbar.update(dt)

                # The Gillespie algorithm randomly samples a possible event
                # with probabilities proportional to the rates
                p = self.rng.uniform(0, rate)

                # Zero-molecular reaction happening
                if p <= np.sum(gen_rates):
                    for reac, rate in zip(self.system.reactions_gen, gen_rates):
                        if p >= rate:
                            p -= rate
                            continue

                        event = self.perform_gen_reaction(reac)
                        self.events.append((self.t, event))
                        break

                # Unimolecular reaction happening
                elif p <= np.sum(gen_rates) + np.sum(uni_rates): 
                    p -= np.sum(gen_rates)
                    
                    for reac in self.system.reactions_uni:
                        if p >= reac.rate * self.counts[reac.spec]:
                            p -= reac.rate * self.counts[reac.spec]
                            continue

                        event = self.perform_uni_reaction(reac)
                        self.events.append((self.t, event))
                        break

                # Bimolecular reaction happening
                else:
                    p -= np.sum(gen_rates) + np.sum(uni_rates)
                    
                    for reac, rates in zip(self.system.reactions_bi, bi_rates):
                        if p >= np.sum(rates):
                            p -= np.sum(rates)
                            continue
                            
                        event = self.perform_bi_reaction(reac, rates)
                        self.events.append((self.t, event))
                        break
           
            # This set sprogress bar value to t0 + tmax
            pbar.update(dt - (self.t - t0 - tmax))
            self.t = t0 + tmax
    
    ### UTILITY FUNCTIONS ###
    def update_count_species(self, **kwargs):
        return self.count_species(**kwargs)
    
    def get_dist(self, t_max=None):
        """ 
            Return the distribution of particle numbers over the lifetime of the system
            using time-averaging.
        """
        counts = np.zeros((self.system.n_species, len(self.events) + 1), dtype=int)
        weights = np.zeros(len(self.events) + 1)

        if t_max is None:
            t_max = self.t
            
        t_last = 0
        i = 0
        
        for t, e in self.events:
            if e[0] == "jump" or t > t_max:
                continue

            weights[i] = t - t_last
            t_last = t

            i += 1
            counts[:,i] = counts[:,i-1]
            if e[0] == "gen":
                product_log = e[2]
                for spec_product in product_log:
                    counts[spec_product][i] += 1
            if e[0] == "uni":
                reac = e[1]
                counts[reac.spec][i] -= 1

                product_log = e[2]
                for spec_product in product_log:
                    counts[spec_product][i] += 1
            elif e[0] == "bi":
                reac = e[1]

                counts[reac.specA][i] -= 1
                counts[reac.specB][i] -= 1

                product_log = e[2]
                for spec_product in product_log:
                    counts[spec_product][i] += 1

        weights[i] = t_max - t_last

        return pdist.ParticleDistribution(counts[:,:i+1], weights=weights[:i+1])

File: cme.jl
"""

This file is a Julia version of the file cme.py; most of it
is a straightforward translation of the original Python code,
optimized for speed using static typing. See cme.py for more
details on the implementation and cme_julia.py for the Python
bindings to this file.

"""
using Distributions;
using RandomNumbers;
using Random;
using PyCall;

@enum ReactionType begin
	GenReactionType = 0
	UniReactionType = 1
	BiReactionType = 2
end

const ProductType = Union{Int64,Tuple{Float64,Int64}}

# All reactions are instances of this struct
# The fields specA and specB may not always be 
# defined, eg. specB for unimolecular reactions
struct Reaction
	rate :: Float64
	products :: Array{ProductType,1}
	specA :: Int64
	specB :: Int64
	type :: ReactionType
end

GenReaction(rate::Float64, products=[]) = Reaction(rate, products, -1, -1, GenReactionType);

UniReaction(rate::Float64, spec::Int64, products=[]) = Reaction(rate, products, spec, -1, UniReactionType);

BiReaction(rate::Float64, specA::Int64, specB::Int64, products=[]) = Reaction(rate, products, specA, specB, BiReactionType)

const DummyReaction = GenReaction(0.0)
   
struct ReactionSystem
    n_species :: Int64
    reactions :: Array{Reaction,1}   
    initial_state :: Array{Int64,1} 
    reactions_gen :: Array{Reaction,1}
    reactions_uni :: Array{Reaction,1}
    reactions_bi :: Array{Reaction,1}
end

function ReactionSystem(n_species::Int64, reactions::Array{Reaction,1})
    return ReactionSystem(n_species, reactions, zeros(Int64, n_species))
end

function ReactionSystem(n_species :: Int64, reactions :: Array{Reaction,1}, initial_state::Array{Int64,1})
    reactions_gen = Reaction[]
    reactions_uni = Reaction[]
    reactions_bi = Reaction[]

    for reaction in reactions
		if reaction.type == GenReactionType
            push!(reactions_gen, reaction)
        elseif reaction.type == UniReactionType
            push!(reactions_uni, reaction)
        elseif reaction.type == BiReactionType
            push!(reactions_bi, reaction)
        else
            throw(ArgumentError("Unknown reaction type for '$reaction'"))
        end
    end

    return ReactionSystem(n_species, reactions, initial_state, reactions_gen, reactions_uni, reactions_bi);
end

const EventType = Tuple{String, Reaction, Array{Int64,1}}

mutable struct ParticleSystem
    system :: ReactionSystem
    cells :: Array{Int64,1}

    rng :: MersenneTwister

    t :: Float64
    events :: Array{Tuple{Float64,EventType},1}
end

function ParticleSystem(system::ReactionSystem, seed::Union{Nothing,Int64}=nothing)
    cells = zeros(Int64, system.n_species)

    if seed == nothing
        rng = MersenneTwister()
    else
        rng = MersenneTwister(seed)
    end

    ret = ParticleSystem(system, cells, rng, 0, [])
    add_initial_molecules(ret)

    return ret
end

function add_initial_molecules(psys::ParticleSystem)
    @assert psys.t == 0

    for (specp1::Int64, n_init::Int64) in enumerate(psys.system.initial_state)
        product_log = add_products(psys, fill(specp1-1, n_init))

        event = ("gen", DummyReaction, product_log)
        push!(psys.events, (0.0, event))
    end
end

function add_products(psys::ParticleSystem, raw_products::Array{Int64,1})::Array{Int64,1}
    products = raw_products

    for product in products
        psys.cells[product+1] += 1
    end

    return products
end

function add_products(psys::ParticleSystem, raw_products::Array{ProductType,1})::Array{Int64,1}
    products = expand_products(psys, raw_products)

    for product in products
        psys.cells[product+1] += 1
    end

    return products
end

function expand_products(psys::ParticleSystem, raw_products::Array{ProductType,1})::Array{Int64,1}
    ret = Int64[]

    for prod in raw_products
        if typeof(prod) == Int64
            push!(ret, prod)
        else
            m, spec = prod
            p = 1 / m

            dist = Geometric(p)
            n = rand(psys.rng, dist)
            append!(ret, fill(spec, n))
        end
    end

    return ret
end

function compute_bi_rates(psys::ParticleSystem)
    rates = Float64[length(psys.system.reactions_bi)]

    for (i, reac) in enumerate(psys.system.reactions_bi)
        if reac.specA == reac.specB
            combs = 0.5 * psys.cells[reac.specA+1] * (psys.cells[reac.specA+1] - 1)
        else
            combs = psys.cells[reac.specA+1] * psys.cells[reac.specB+1]
        end
            
        rates[i] = reac.rate * combs
    end

    return rates
end

function perform_gen_reaction(psys::ParticleSystem, reaction::Reaction)
    product_log = add_products(psys, reaction.products)
    return ("gen", reaction, product_log)
end

function perform_uni_reaction(psys::ParticleSystem, reaction::Reaction)
    psys.cells[reaction.specA+1] -= 1
    product_log = add_products(psys, reaction.products)
    return ("uni", reaction, product_log)
end

function perform_bi_reaction(psys::ParticleSystem, reaction::Reaction)
    psys.cells[reaction.specA+1] -= 1
    psys.cells[reaction.specB+1] -= 1
    product_log = add_products(psys, reaction.products)
    return ("bi", reaction, product_log)
end
        
mutable struct PBarWrapper
    pbar_raw::PyObject
    buf::Float64
    tmax::Float64
end

function pbar_update_wrapper(pbar_update::PyObject, pbar::PBarWrapper, dt::Float64)
    if pbar.buf + dt >= pbar.tmax / 100.0
        pycall(pbar_update, Any, pbar.pbar_raw, pbar.buf+dt)
        pbar.buf = 0
    else
        pbar.buf += dt
    end
end

function run(psys::ParticleSystem, tmax::Float64; disable_pbar::Bool=true,
             pbar_create = nothing, pbar_update = nothing, pbar_close = nothing)
    t0 = psys.t

    gen_rates = [ reac.rate for reac in psys.system.reactions_gen ]

    pbar = missing
    if !disable_pbar && pbar_create !== nothing
        pbar_raw = pycall(pbar_create, Any, tmax, "Time simulated: ", "tu")
        pbar = PBarWrapper(pbar_raw, 0.0, tmax)
    end

    dt = 0
    while true    
        uni_rates = [ reac.rate * psys.cells[reac.specA+1] for reac in psys.system.reactions_uni ]
        bi_rates = compute_bi_rates(psys)
        rate = sum(gen_rates) + sum(uni_rates) + sum(bi_rates)

        if rate == 0.0 || !isfinite(rate)
            dt = 0
            break
        end

        dt = randexp(psys.rng, Float64) / rate
        psys.t += dt

        if psys.t >= t0 + tmax
            break
        end

        if !ismissing(pbar)
            pbar_update_wrapper(pbar_update, pbar, dt)
        end

        p = rand(psys.rng, Float64) * rate

        if p <= sum(gen_rates)
            for reac in psys.system.reactions_gen
                if p >= reac.rate
                    p -= reac.rate
                    continue
                end

                event = perform_gen_reaction(psys.reac)
                push!(psys.events, (psys.t, event))
                break
            end
        elseif p <= sum(gen_rates) + sum(uni_rates)
            p -= sum(gen_rates)
            
            for reac in psys.system.reactions_uni
                if p >= reac.rate * psys.cells[reac.specA+1]
                    p -= reac.rate * psys.cells[reac.specA+1]
                    continue
                end

                event = perform_uni_reaction(psys, reac)
                push!(psys.events, (psys.t, event))
                break
            end
        else
            p -= (rate - sum(bi_rates))
            
            for (reac, rates) in zip(psys.system.reactions_bi, bi_rates)
                if p >= sum(rates)
                    p -= sum(rates)
                    continue
                end
                    
                event = perform_bi_reaction(psys, reac)
                push!(psys.events, (psys.t, event))
                break
            end
        end

    end

    if !ismissing(pbar)
        pycall(pbar_update, Any, pbar.pbar_raw, pbar.buf + dt - (psys.t - t0 - tmax))
        pycall(pbar_close, Any, pbar.pbar_raw)
    end

    psys.t = t0 + tmax
end
         
function get_dist_data(psys::ParticleSystem)#::Array{Float64}
    return get_dist_data(psys,
                         zeros(Float64, zeros(Int64, psys.system.n_species)...),
                         zeros(Int64, psys.system.n_species),
                         0.0, 
                         0)
end

function get_dist_data(psys::ParticleSystem, 
                       data_old::Array{Float64}, cells_old::Array{Int64,1}, t_min::Float64, n_old_events::Int64)#::Array{Float64}
    counts = Array{Int64,2}(undef, length(psys.events) - n_old_events + 1, psys.system.n_species)
    weights = Array{Float64,1}(undef, length(psys.events) - n_old_events + 1)

    counts[1,:] = cells_old
    
    t_last :: Float64 = t_min
    t :: Float64 = t_min

    i ::Int64 = 1

	reac::Reaction = DummyReaction

    for (t, ev) in psys.events[n_old_events+1:end]
        weights[i] = t - t_last
        t_last = t

        i += 1
        counts[i,:] = counts[i-1,:]
        if ev[1] == "gen"
            product_log = ev[3]
            for spec_product in product_log
                counts[i,spec_product+1] += 1
            end
        elseif ev[1] == "uni"
            reac = ev[2]
            counts[i,reac.specA+1] -= 1

            product_log = ev[3]
            for spec_product in product_log
                counts[i,spec_product+1] += 1
            end
        elseif ev[1] == "bi"
            reac = ev[2]

            counts[i,reac.specA+1] -= 1
            counts[i,reac.specB+1] -= 1

            product_log = ev[3]
            for spec_product in product_log
                counts[i,spec_product+1] += 1
            end
        end
    end

    @assert minimum(counts) >= 0

    weights[end] = psys.t - t_last

    return convert_dist_data(counts, weights, data_old, t_min)
end

function convert_dist_data(counts::Array{Int64,2}, weights::Array{Float64})::Array{Float64}
    return convert_dist_data(counts, weights, zeros(Int64, 1, size(pseudocounts, 2)))
end


function convert_dist_data(counts::Array{Int64,2}, weights::Array{Float64},
                           data_old::Array{Float64}, weight_old::Float64=0)::Array{Float64}
    bounds_old::Tuple{Vararg{Int64}} = size(data_old)
    bounds_new::Array{Int64,1} = [ maximum(counts[:,i]) + 1 for i in 1:size(counts, 2) ]
    @assert length(bounds_old) == length(bounds_new)

    bounds::Array{Int64,1} = [ max(bo, bn) for (bo, bn) in zip(bounds_old, bounds_new) ]
    data_flat::Array{Float64,1} = zeros(Float64, prod(bounds))
	data::Array{Float64} = reshape(data_flat, bounds...)
    
    slices::Array{UnitRange{Int64}} = [ 1:bo for bo in bounds_old ]

	data_strides::Array{Int64,1} = [ s for s in strides(data) ]

    for idx::Tuple{Vararg{Int64}} in Iterators.product(slices...)
		idx_cart = CartesianIndex(idx)
        data[idx_cart] += weight_old * data_old[idx_cart]
    end

    for (i, wt) in enumerate(weights)
		idx_flat = sum(counts[i,:] .* data_strides) + 1
        data_flat[idx_flat] += wt
    end
            
    data_flat ./= sum(data_flat)

    return data
end

